_source/_posts/2020-07-02-spring-boot-using-java-modules.md

```diff
@@ -0,0 +1,465 @@
+
+## Project Structure
+
+In this article, we'll be covering how to develop a simple application with two modules: the `application` module that contains the web-facing classes and the `persistence` module that contains the data access layer. We'll also be using a couple of dependencies to illustrate how to use those on a modular application: `spring-boot-starter-data-mongodb` and `okta-spring-boot-starter`
+
+The project source code can be found at [GitHub](https://github.com/brunocleite/spring-boot-using-java-modules)
+
+### How to Structure a Modular Project with Maven?
+
+Let's create this project folder structure manually to better understand how it should be structured. Each module will live inside a separate directory and have it's own `pom.xml` file. There will also be a `pom.xml` on the project root that will serve as the parent pom for the modules.  You can create the following folder structure:
+
+```
+.
+├── application
+│   ├── pom.xml
+│   └── src
+│       └── main
+│           ├── java
+│           └── resources
+├── persistence
+│   └── src
+│       └── main
+│           └── java
+└── pom.xml
+```
+
+### How to Tie Up the Three pom.xml Files with Maven?
+
+First, let's define the root `pom.xml`. It will contain the common `<parent>` indication to `spring-boot-started-parent` and two entries on `<module>` section, that are the name of the directories for the modules we are developing. Please note that those are specific to Maven and specify sub-projects and have nothing to do with the Java modules that we'll be working on later.
+
+```xml
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-starter-parent</artifactId>
+        <version>2.3.1.RELEASE</version>
+        <relativePath/> <!-- lookup parent from repository -->
+    </parent>
+    <groupId>com.okta.developer</groupId>
+    <artifactId>spring-boot-with-modules</artifactId>
+    <version>0.0.1-SNAPSHOT</version>
+    <packaging>pom</packaging>
+
+    <properties>
+    	<java.version>11</java.version>
+    </properties>
+
+    <modules>
+        <module>application</module>
+        <module>persistence</module>
+    </modules>
+
+</project>
+```
+
+The `application` module will have a pom.xml like below, pointing to the parent pom.xml that was described above. It will also have a dependency on `spring-boot-starter-web` because we'll be creating some REST endpoints on it and also a dependency on our 'persistence' module that will be described next.
```

Comment:
```suggestion
The `application` module will have a `pom.xml` like below, pointing to the parent pom.xml that was described above. It will also have a dependency on `spring-boot-starter-web` because we'll be creating some REST endpoints on it and also a dependency on our 'persistence' module that will be described next.
```